kube-runtime/src/coordination.rs

```diff
@@ -0,0 +1,603 @@
+//! Coordination utilities built around the `coordination.k8s.io/v1` API.
+//!
+//! This implementation uses only Kubernetes `coordination.k8s.io/v1/Lease` objects for
+//! coordination. Every client running this `LeaderElector` task maintains a watch on the
+//! lease object, and will receive updates as the lease holder heartbeats the lease.
+//!
+//! Applications using this API can use the spawn handle returned from spawning the
+//! `LeaderElector` task to monitor lease state and to govern application behavior based on that
+//! state. E.G.,
+//!
+//! ```rust,ignore
+//! // Spawn a leader elector task, and get a handle to the state channel.
+//! let handle = LeaderElector::spawn(/* ... */);
+//! let state_chan = handle.state();
+//!
+//! // Before taking action as a leader, just check the channel to ensure
+//! // the lease is currently held by this process.
+//! if state_chan.borrow().is_leader() {
+//!     // Only perform leader actions if in leader state.
+//! }
+//!
+//! // Or, for a more sophisticated pattern, watch the state channel for changes,
+//! // and use it to drive your application's state machine.
+//! let state_stream = tokio_stream::wrappers::WatchStream::new(state_chan);
+//! loop {
+//!     tokio::select! {
+//!         Some(state) = state_stream.next() => match state {
+//!             LeaderState::Leader => (), // Leader tasks.
+//!             _ => (), // Non-leader tasks.
+//!         },
+//!     }
+//! }
```

Comment:
So, I've been using this code for a long time now, and the pattern that I've found most useful is actually to use the `watch::Receiver<LeaderState>` directly.

One of the good things about this type is that it can be used to directly access the current value: `.borrow()`. It can also be used in a `select!` with `.changed()`. This serves to reduce the amount of additional code needed to work with this data. Don't need to retain a copy of the most recent value elsewhere, and so on.

We could create a type wrapper around this which exposes both of these features. The problem with using `tokio_stream::wrappers::WatchReceiver` is that it does not expose a `.borrow()` for directly checking that last value.